tokio/tests/sync_once_cell.rs

```diff
@@ -0,0 +1,97 @@
+#![warn(rust_2018_idioms)]
+#![cfg(feature = "full")]
+
+use tokio::runtime::Runtime;
+use tokio::sync::{AlreadyInitializedError, Lazy, NotInitializedError, OnceCell};
+use tokio::time::sleep;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::time::{Duration, Instant};
+
+fn func1() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+    async fn f() -> u32 {
+        5
+    }
+    Box::pin(f())
+}
+
+fn func2() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+    async fn f() -> u32 {
+        sleep(Duration::from_secs(1)).await;
+        10
+    }
+    Box::pin(f())
+}
+
+#[test]
+fn get_or_init() {
+    let rt = Runtime::new().unwrap();
+
+    static ONCE: OnceCell<u32> = OnceCell::new();
+
+    rt.block_on(async {
+        let result1 = rt
+            .spawn(async { ONCE.get_or_init(func1).await })
+            .await
+            .unwrap();
+
+        let result2 = rt
+            .spawn(async { ONCE.get_or_init(func2).await })
+            .await
+            .unwrap();
+
+        assert_eq!(*result1, 5);
+        assert_eq!(*result2, 5);
+    });
+}
+
+#[test]
+fn set_and_get() {
+    let rt = Runtime::new().unwrap();
+
+    static ONCE: OnceCell<u32> = OnceCell::new();
+
+    rt.block_on(async {
+        let _ = rt.spawn(async { ONCE.set(5) }).await;
+        let value = ONCE.get().unwrap();
+        assert_eq!(*value, 5);
+    });
+}
+
+#[test]
+fn get_uninit() {
+    static ONCE: OnceCell<u32> = OnceCell::new();
+    let uninit = ONCE.get();
+    assert_eq!(uninit, Err(NotInitializedError));
+}
+
+#[test]
+fn set_twice() {
+    static ONCE: OnceCell<u32> = OnceCell::new();
+
+    let first = ONCE.set(5);
+    assert_eq!(first, Ok(()));
+    let second = ONCE.set(6);
+    assert_eq!(second, Err(AlreadyInitializedError));
+}
+
+#[test]
+fn lazy() {
+    // func2 sleeps for 1 second
+    static LAZY: Lazy<u32> = Lazy::new(func2);
+    let rt = Runtime::new().unwrap();
```

Comment:
Can you use the pause time feature to make this test run faster?

Reply:
I'm not sure whether I misunderstand you here. But if we pause the timer, how is the test supposed to pass?

Reply:
Afaik the timer automatically advances if things are only waiting for timers. Alternatively please change the durations to something like 10ms so it doesn't take a full second to run them.

Reply:
Yes, 1 second was a conservative estimate. Could the time the runtime takes to execute (or more precisely if the second task is executed before the first one) not lead to the task failing?

Reply:
Generally, avoiding this kind of thing is the purpose of the time-pausing utilities.

Reply:
It wasn't clear to me how to use `time::pause` to avoid this problem. `time::pause` isn't used in any tests in the code base besides those that test the timer functionality itself.